### PR TITLE
Composer object store download retry

### PR DESCRIPTION
**Pull request by bigning**
# What does this PR do?

add object_store checkpoint downloading retry

# test
bigning-debug-cp-loading-2-J6z0wB  **ConnectionError**
in the logging, there is one rank succeeded at 2nd attempt. 
image

another run: bigning-debug-cp-loading-2-9Lq8am **ChunkedEncodingError**
image

### CODE REVIEW

**Review by b-chu — CHANGES_REQUESTED** [review]
Some failing tests but overall looks good. Let's clean up the unused import and pyright ignores. This would retry with every exception though, not just the ones that aren't retried from the object store. So you would 5x the number of retries that were already happening.

**Comment by bigning** [comment]
>So you would 5x the number of retries that were already happening.

@b-chu , for the retries that already happening, if the oci internal retry works, then we don't do extra retry. Only if the internal retry fails, we do extra retry (i think that's kind of what we want. :) ).

File: composer/utils/checkpoint.py [hunk 1]
```diff
@@ -188,6 +190,24 @@ def read_metadata(self) -> Metadata:
         return super().read_metadata()
 
 
+@retry(num_attempts=5)
+def download_object_or_file(
+    object_name: str,
+    file_destination: Union[str, pathlib.Path],
```
Comment (b-chu) [inline composer/utils/checkpoint.py:196]: we already import Path

File: composer/utils/checkpoint.py [hunk 2]
```diff
@@ -188,6 +190,24 @@ def read_metadata(self) -> Metadata:
...
+):
+    if isinstance(object_store, ObjectStore):
+        object_store.download_object(
+            object_name=object_name,
+            filename=file_destination,
+        )
+    else:
+        object_store.download_file(
+            remote_file_name=object_name,
+            destination=cast(str, file_destination),
```
Comment (b-chu) [inline composer/utils/checkpoint.py:207]: does `str(file_destination)` work?

File: composer/utils/checkpoint.py [hunk 3]
```diff
@@ -228,7 +248,8 @@ def __init__(
     def read_data(self, plan: LoadPlan, planner: LoadPlanner):
         # Download files if not using HSDP or if on first replica with HSDP enabled
         first_replica = self.device_mesh is None or self.device_mesh.ndim == 1 or (
-            self.device_mesh.ndim >= 2 and self.device_mesh.get_local_rank(mesh_dim=0) == 0
+            self.device_mesh.ndim >= 2 and
+            self.device_mesh.get_local_rank(mesh_dim=0) == 0  # pyright: ignore[reportGeneralTypeIssues]
```
Comment (b-chu) [inline composer/utils/checkpoint.py:252]: Upgrading your local env should fix the pyright issues

**Review by b-chu — COMMENTED** [review]
Couple of minor comments

**Review by b-chu — APPROVED** [review]
Thanks for the cleanup!

File: composer/utils/checkpoint.py [hunk 4]
```diff
@@ -188,6 +189,24 @@ def read_metadata(self) -> Metadata:
...
+    if isinstance(object_store, ObjectStore):
+        object_store.download_object(
+            object_name=object_name,
+            filename=file_destination,
+        )
+    else:
+        object_store.download_file(
+            remote_file_name=object_name,
+            destination=str(file_destination),
+        )
```
Comment (mvpatel2000) [inline composer/utils/checkpoint.py:207]: can we just add `download_object` as a wrapper fn to LoggerDestination and forward args to `download_file`? 

@eracah thoughts?

Comment (bigning) [reply composer/utils/checkpoint.py:207]: @mvpatel2000  I want to minimize this PR. Can we do that in a follow-up PR if we decide to do so ?

Comment (bigning) [reply composer/utils/checkpoint.py:207]: discussed offline with @mvpatel2000 , not a blocker for this PR.

**Pull request merged**